devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
```

Comment:
>queues requests when there are no jobs [](http://example.com/codeflow?start=118&length=38)

Maybe my real question is, what is a "job" and what is a "request"?

Reply:
A "job" is what a producer puts in the queue. A "request" is what a consumer puts in the queue. 

In GeoReplication terms, `job_queue_push` is called when the state provider calls `Begin_Replicate`. 

Each replication stream has its own `JOB_QUEUE_SESSION`. Secondary state providers call replication_stream.get_next() which results in a `job_queue_session_get()` call. After receiving the operation, secondary state providers call Acknowledge(), which results in a `job_queue_session_pop()` call. Once enough sessions have popped the operation ("enough" is configurable") then the `JOB_COMPLETE_CALLBACK` is called and the GeoReplicator indicates to the primary state provider that the operation has been replicated.